homeassistant/components/binary_sensor/w800rf32.py

```diff
@@ -0,0 +1,179 @@
+        })
+    },
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Binary Sensor platform to w800rf32."""
+    import W800rf32 as w800rf32mod
+    binary_sensors = []
+    # device_id --> "c1 or a3" X10 device. entity (type dictionary)
+    # --> name, device_class etc
+    for device_id, entity in config[CONF_DEVICES].items():
+
+        if device_id in hass.data[DOMAIN]['entities']:
+            continue
+
+        _LOGGER.debug("Add %s w800rf32.binary_sensor (class %s)",
+                      entity[ATTR_NAME], entity.get(CONF_DEVICE_CLASS))
+
+        device = W800rf32BinarySensor(
+            device_id, entity.get(CONF_NAME), entity.get(CONF_DEVICE_CLASS),
+            entity[CONF_FIRE_EVENT], entity.get(CONF_OFF_DELAY))
+
+        hass.data[DOMAIN]['entities'][device_id] = device
+        binary_sensors.append(device)
+
+    add_entities(binary_sensors)
+
+    def binary_sensor_update(event):
+        """Call for control updates from the w800rf32 gateway."""
+        sensor = None
+
+        if not isinstance(event, w800rf32mod.W800rf32Event):
+            return
+
+        dev_id = event.device.lower()
+        command = event.command
+
+        #
+        # # get the name, ex: motion_hall
+        if dev_id in hass.data[DOMAIN]['entities']:
+            sensor = hass.data[DOMAIN]['entities'][dev_id]
+
+        if sensor is None:
+            return
+
+        if not isinstance(sensor, W800rf32BinarySensor):
+            return
+
+        _LOGGER.debug(
+            "BinarySensor update (Device ID: %s Command %s ...)",
+            dev_id, command)
+
+        # Update the w800rf32 device state
+        if command in ('On', 'Off'):
+            is_on = command == 'On'
+            sensor.update_state(is_on)
+
+        # Fire event
+        if sensor.should_fire_event:
+            sensor.hass.bus.fire(
+                EVENT_BUTTON_PRESSED, {
+                    ATTR_ENTITY_ID:
+                        sensor.entity_id,
+                    ATTR_STATE: command.lower()
+                }
+            )
+            _LOGGER.debug(
+                "w800rf32 fired event: (event_type: %s, %s: %s, %s: %s)",
+                EVENT_BUTTON_PRESSED,
+                ATTR_ENTITY_ID,
+                sensor.entity_id,
+                ATTR_STATE,
+                command.lower()
+            )
+
+        if (sensor.is_on and sensor.off_delay is not None and
+                sensor.delay_listener is None):
+
+            def off_delay_listener(now):
+                """Switch device off after a delay."""
+                sensor.delay_listener = None
+                sensor.update_state(False)
+
+            sensor.delay_listener = evt.track_point_in_time(
+                hass, off_delay_listener, dt_util.utcnow() + sensor.off_delay)
+
+    # Subscribe to main w800rf32 events
+    dispatcher_connect(hass, DEVICE_TYPE, binary_sensor_update)
```

Comment:
Move this to entity coroutine method `async_added_to_hass`.

Reply:
async_added_to_hass?  will that get called automatically when I create an instance of my entity class?  Where does it get called from?  I looked at the example you gave above but it doesn't get called from the platform code.

Reply:
I haven't done any python async stuff yet.  Once I get this working I was going to change this code to an async version, one step at a time.  Thanks for your patience.

Reply:
Never mind about the async_added_to_hass, I figured it out.